art/detection/subsetscanning/detector.py

```diff
@@ -0,0 +1,184 @@
+
+logger = logging.getLogger(__name__)
+
+
+class SubsetScanningDetector(Classifier):
+    """ Fast generalized subset scan based detector"""
+    def __init__(self, classifier, bgd_data, layer):
+        super(SubsetScanningDetector, self).__init__(clip_values=classifier.clip_values,
+                                                     channel_index=classifier.channel_index,
+                                                     defences=classifier.defences,
+                                                     preprocessing=classifier.preprocessing)
+        self.classifier = classifier
+        self.bgd_data = bgd_data
+
+        # Ensure that layer is well-defined:
+        if isinstance(layer, six.string_types):
+            if layer not in classifier.layer_names:
+                raise ValueError('Layer name %s is not part of the graph.' % layer)
+            self._layer_name = layer
+        elif isinstance(layer, int):
+            if layer < 0 or layer >= len(classifier.layer_names):
+                raise ValueError('Layer index %d is outside of range (0 to %d included).'
+                                 % (layer, len(classifier.layer_names) - 1))
+            self._layer_name = classifier.layer_names[layer]
+        else:
+            raise TypeError('Layer must be of type `str` or `int`.')
+
+        bgd_activations = classifier.get_activations(bgd_data, self._layer_name)
+        if len(bgd_activations.shape) == 4:
+            dim2 = bgd_activations.shape[1] * bgd_activations.shape[2] * bgd_activations.shape[3]
+            bgd_activations = np.reshape(bgd_activations, (bgd_activations.shape[0], dim2))
+
+        self.sorted_bgd_activations = np.sort(bgd_activations, axis=0)
+
+
+    def calculate_pvalue_ranges(self, x):
+
+        bgd_activations = self.sorted_bgd_activations
+        eval_activations = self.classifier.get_activations(x, self._layer_name)
+
+        if len(eval_activations.shape) == 4:
+            dim2 = eval_activations.shape[1] * eval_activations.shape[2] * eval_activations.shape[3]
+            eval_activations = np.reshape(eval_activations, (eval_activations.shape[0], dim2))
+
+        bgrecords_n = bgd_activations.shape[0]
+        records_n = eval_activations.shape[0]
+        atrr_n = eval_activations.shape[1]
+
+        pvalue_ranges = np.empty((records_n, atrr_n, 2))
+
+        for j in range(atrr_n):
+
+            pvalue_ranges[:, j, 0] = np.searchsorted(bgd_activations[:, j], eval_activations[:, j], side='right')
+            pvalue_ranges[:, j, 1] = np.searchsorted(bgd_activations[:, j], eval_activations[:, j], side='left')
+
+        pvalue_ranges = bgrecords_n - pvalue_ranges
+
+        pvalue_ranges[:, :, 0] = np.divide(pvalue_ranges[:, :, 0], bgrecords_n+1)
+        pvalue_ranges[:, :, 1] = np.divide(pvalue_ranges[:, :, 1] + 1, bgrecords_n+1)
+
+        return pvalue_ranges
+
+    def scan(self, clean_x, adv_x, cleanssize=None, advssize=None, run=10):
```

Comment:
Could you please add a docstring describing what the function does, its arguments and return similar to the other function docstrings in ART?